### PR TITLE
31 add react router

### LINKED ISSUE #31: Add React Routing

### PR DESCRIPTION
**Pull request by rafibarash**
Fixes #31. 

Touched a lot of files here, so feel free to make comments if there are issues you see, or unnecessary changes I made

## Changes made

- Added React Router to project, and made a new directory for `views`, to handle page components and components that are found solely within that page.
- Left shared components in the `components` directory.  
- Supported routes for `/` homepage, `/tickets` where it simply lists recent tickets and suggest selecting a ticket, and `/tickets/{id}` to show fields associated with a specific ticket
- Changed page header (the "Vigilante Web Heist" header) into a header link to go back home (`/`)
- Got ahead of this PR and created a helper function in `utils/api.js` for fetching a ticket by ID. I am thinking we can abstract api calls and keep their logic in here

## Other notes

- Slightly refactored the app level state to contain a `curTicket`, and the response in the `UserInput` to set that ticket. I am planning on refactoring this to take away the `curTicket` from the app level state, and instead simply redirect the user to the new ticket's route after submitting inside `UserInput`. The `Ticket` component that handles this route will always load its ticket's resources after mounting, similar to how the `RecentTickets` component fetches data from the backend. This fixes issues that would arise when navigating to a specific ticket and not having that ticket's data in the app level state

### CODE REVIEW

**Comment by williamtdr** [comment]
Looks good! Going to download and take a look....

**Comment by williamtdr** [comment]
The redirect feature isn't working for me. This is what I see on the console:
image

File: src/views/Home/Home.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import React, { Component } from 'react';
```
Comment (williamtdr) [inline src/views/Home/Home.js]: nit: directory names should probably be lowercase

File: src/views/Tickets/Tickets.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { Header } from 'semantic-ui-react';
```
Comment (williamtdr) [inline src/views/Tickets/Tickets.js]: unused

File: src/views/Tickets/Tickets.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import { useRouteMatch } from 'react-router-dom';
+
+// Routes
+import NestedTicketRoutes from '../../routes/NestedTicketRoutes';
+
+const Tickets = ({ ticket }) => {
+  const { path, url } = useRouteMatch();
```
Comment (williamtdr) [inline src/views/Tickets/Tickets.js]: `url` is unused, suggested: `const { path } = useRouteMatch();`

File: src/routes/NestedTicketRoutes.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// View sub-components
+import Ticket from '../views/Tickets/Ticket';
+
+const NestedTicketRoutes = ({ path, ticket }) => (
+  <Switch>
+    <Route exact path={path}>
+      <Header as="h5">Please select a ticket.</Header>
+      <RecentTickets />
+    </Route>
+    <Route path={`${path}/:ticketID`}>
```
Comment (williamtdr) [inline src/routes/NestedTicketRoutes.js]: really like this syntax!

File: src/utils/api.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+export const fetchTicket = async ticketID => {
+  try {
+    const res = await fetch(`/api/tickets/${ticketID}`, {
+      method: 'GET',
+    });
+    return await res.json();
```
Comment (williamtdr) [inline src/utils/api.js]: don't need to `await` a synchronous call. consider promise chaining.

File: src/utils/api.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+  } catch (error) {
+    console.error(error);
```
Comment (williamtdr) [inline src/utils/api.js]: nit: explicitly return null on failure

**Comment by rafibarash** [comment]
@williamtdr I'm adding the redirect feature in another PR, should have it up tomorrow. And sounds good I'll also look at your comments tomorrow

File: src/utils/api.js (hunk 5 above)
Comment (rafibarash) [reply src/utils/api.js]: the res.json() is asyncronous, are you talking about that or something else?

screenshot from the [fetch documentation](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)
Screen Shot 2019-11-12 at 7 50 47 PM

**Pull request merged**